### PR TITLE
Trigger plugin

### PR DESCRIPTION
**Pull request by civing**
Adds triggers to be used when detecting a certain string and starting a trigger session. A trigger session begins when the magic string sequence is inserted, for instance an @ and will keep on sending user input back to the consumer until certain stop conditions occur.

fixes: Lundalogik/crm-feature#4430
Co-authored-by: John Traas <john.traas@lime.tech>

## Review:
- [ ] Commits are [atomic](https://seesparkbox.com/foundry/atomic_commits_with_git)
- [ ] Commits have the correct *type* for the changes made
- [ ] Commits with *breaking changes* are marked as such

### Browsers tested:
(Check any that applies, it's ok to leave boxes unchecked if testing something didn't seem relevant.)

Windows:
- [ ] Chrome
- [ ] Edge
- [ ] Firefox

Linux:
- [ ] Chrome
- [ ] Firefox

macOS:
- [ ] Chrome
- [ ] Firefox
- [ ] Safari

Mobile:
- [ ] Chrome on Android
- [ ] iOS

### CODE REVIEW

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    | ')'
+    | "'";
+
+export type CustomElementInstance = Omit<CustomElement, 'attributes'> & {
+    attributes: Record<string, any>;
+};
+
+export type TextEditorNode = {
+    node: CustomElementInstance | string;
+    childs?: Array<TextEditorNode | string>;
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:49]:
```suggestion
    children?: Array<TextEditorNode | string>;
```
:sweat_smile:

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+};
+
+export interface TextEditorInserter {
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:52]: Maybe this should just be named `TextEditor` if we want it to be able to add more methods to it in the future

```suggestion
export interface TextEditor {
```

File: src/components/text-editor/text-editor.tsx [hunk 3]
```diff
@@ -105,6 +110,9 @@ export class TextEditor implements FormComponent<string> {
     @Prop()
     public plugins: CustomElement[] = [];
```
Comment (jgroth) [inline src/components/text-editor/text-editor.tsx:111]: Not part of this PR, but I think we should rename this to something like `customElements` instead. Calling it `plugins` is a bit confusing for the consumer

File: src/components/text-editor/text-editor.tsx [hunk 4]
```diff
@@ -142,6 +150,15 @@ export class TextEditor implements FormComponent<string> {
...
     public change: EventEmitter<string>;
 
+    @Event()
+    public triggerStart: EventEmitter<TextEditorInserter>;
+
+    @Event()
+    public triggerStop: EventEmitter<void>;
+
+    @Event()
+    public triggerChange: EventEmitter<string>;
```
Comment (jgroth) [inline src/components/text-editor/text-editor.tsx:160]: I think these should be a bit more symmetric, now all of them have different event details. All of them should probably include the `TextEditor` somehow, maybe we need to wrap it in something like this
```ts
interface TriggerEventDetail {
    /**
     * The trigger that triggered this event
     */
    trigger: string;

    /**
     * The text editor
     */
    textEditor: TextEditor;

    /**
     * Current value of the trigger
     */
    value: string
}
```

`trigger` is needed in case you have multiple trigger characters, otherwise you would not know what trigger emitted the event.

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    | ']'
+    | '{'
+    | '}'
+    | '('
+    | ')'
+    | "'";
+
+export type CustomElementInstance = Omit<CustomElement, 'attributes'> & {
+    attributes: Record<string, any>;
+};
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:45]: Why do we need to re-type this? :thinking:

File: src/components/text-editor/text-editor.tsx (hunk 3 above)
Comment (civing) [reply src/components/text-editor/text-editor.tsx:111]: 100% agree. This is the plan as well

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 5 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:45]: In defining a `CustomElement` you will have the attributes as a list of strings. When you are actually "instantiating" it you want the attributes as a record. :)

File: src/components/text-editor/text-editor.tsx (hunk 4 above)
Comment (civing) [reply src/components/text-editor/text-editor.tsx:160]: Yes! I like this a lot.

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 2 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:52]: The name is maybe a bit generic?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 1 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: Fixed in latest commits

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 2 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: Fixed in latest commit

File: src/components/text-editor/text-editor.tsx (hunk 4 above)
Comment (civing) [reply src/components/text-editor/text-editor.tsx]: Fixed in latest commit

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+     */
+    textEditor: TextEditor;
+
+    /**
+     * Current value of the trigger
+     */
+    value: string;
+}
+
+export const inserterFactory = (view: EditorView): TextEditor => {
```
Comment (hannahu) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:73]: Have we investigated if there are any existing trigger functionality we can import instead of re-inventing the wheel?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+    view.dom.dispatchEvent(event);
+};
+
+export const createTriggerPlugin = (triggerCharacters: TriggerCharacter[]) => {
+    let activeTrigger: TriggerCharacter | null = null;
+    let triggerText = '';
+    let pluginView: EditorView | null = null;
+    let triggerPosition: number | null = null;
+
+    return new Plugin({
```
Comment (hannahu) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:229]: Have we investigated if there are any existing trigger functionality we can import instead of re-inventing the wheel?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 2 above)
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: Will this possibly become a bit confusing? We have the `text-editor` in lime-elements, and will have the `limebb-text-editor` as well (though perhaps this limebb component should be named differently).

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 7 above)
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:229]: There's no bundled trigger functionality with ProseMirror. There are two lightweight libraries out there that implement very basic trigger functionality, but they weren't robust enough and didn't suit our needs.

File: src/components/text-editor/examples/text-editor-custom-triggers.tsx [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+ */
+@Component({
+    tag: 'limel-example-text-editor-triggers',
+    shadow: true,
+    styleUrl: 'text-editor-custom-triggers.scss',
+})
+export class TextEditorCustomTriggersExample {
+    constructor() {
+        this.portalId = createRandomString();
+        this.globalClickListener = this.globalClickListener.bind(this);
```
Comment (jgroth) [inline src/components/text-editor/examples/text-editor-custom-triggers.tsx:44]: We don't need explicit binding like this if we use an arrow function instead

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+) => {
+    const event = new CustomEvent<TriggerEventDetail>(type, {
+        detail: getTriggerEventDetail(view, trigger, value),
+        bubbles: true,
+        composed: true,
+    });
+    view.dom.dispatchEvent(event);
+};
+
+export const createTriggerPlugin = (triggerCharacters: TriggerCharacter[]) => {
```
[the remark this answers is not in the document]
Comment (jgroth) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:223]: Yes, please split it up :smile:

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 5 above)
Comment (jgroth) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:45]: I see. I don't like the name `CustomElementInstance` :smile: Why does it need the `Instance` suffix?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+
+import { Node, NodeType, Schema } from 'prosemirror-model';
+import { Plugin, PluginKey, EditorState } from 'prosemirror-state';
+import { ReplaceStep, ReplaceAroundStep } from 'prosemirror-transform';
+import { EditorView } from 'prosemirror-view';
+import { CustomElement } from 'src/interface';
+
+export const triggerPluginKey = new PluginKey('triggerPlugin');
+
+export type TriggerCharacter =
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:11]: What did we say about this? Should we still keep it or do we just have to keep it for now?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 11]
```diff
@@ -0,0 +1,352 @@
+// import { ListItem } from 'src/components/list/list-item.types';
+
+import { Node, NodeType, Schema } from 'prosemirror-model';
+import { Plugin, PluginKey, EditorState } from 'prosemirror-state';
+import { ReplaceStep, ReplaceAroundStep } from 'prosemirror-transform';
+import { EditorView } from 'prosemirror-view';
+import { CustomElement } from 'src/interface';
+
+export const triggerPluginKey = new PluginKey('triggerPlugin');
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:9]: Do we need to export this?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 6 above)
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:73]: Why not make this into a class? :slightly_smiling_face: Perhaps move it to a separate file?

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+    trigger: TriggerCharacter,
+    value: string,
+) => {
+    const event = new CustomEvent<TriggerEventDetail>(type, {
+        detail: getTriggerEventDetail(view, trigger, value),
+        bubbles: true,
+        composed: true,
+    });
+    view.dom.dispatchEvent(event);
+};
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:221]: This might be fine, but I think it would be better if we emit the events from the component instead. We can have a callback on the component that we invoke from here that emits the events, e.g.

```ts
class TextEditor {
    @Event()
    public triggerStart: EventEmitter<TriggerEventDetail>;

    private emitTriggerStart(data) {
        this.triggerStart.emit(...);
    }
}
```

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 2 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: Yeah, I had the same thought

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 7 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:229]: But wheels are so fun to invent! 🛞

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+                                const toPos = step.to;
+
+                                if (slice && slice.size) {
+                                    // Text added
+                                    textAdded += slice.content.textBetween(
+                                        0,
+                                        slice.size,
+                                    );
+                                } else if (fromPos !== toPos) {
+                                    // Text removed
```
Comment (jgroth) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:317]: image

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 11 above)
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:9]: We need to import it into the prosemirror adapter so that we can instantiate the plugin when the text editor view state is created.

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:282]: See https://github.com/Lundalogik/lime-elements/pull/3268/commits/b50ca82479b1ebda1b610441e2f660ea802b7b27

File: src/components/text-editor/examples/text-editor-custom-triggers.tsx [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+        const items = this.items.filter((item: ListItem<number>) =>
+            item.text.toLowerCase().includes(this.inputText),
+        );
+
+        const dropdownZIndex = getComputedStyle(this.host).getPropertyValue(
+            '--dropdown-z-index',
+        );
+
+        return [
+            <limel-portal
```
[the remark this answers is not in the document]
Comment (civing) [reply src/components/text-editor/examples/text-editor-custom-triggers.tsx:199]: : (

File: src/components/text-editor/examples/text-editor-custom-triggers.tsx [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+ *
+ * In this example we pass either a text or a `limel-chip` representing some chosen user
+ * in a mention like situation.
+ */
+@Component({
+    tag: 'limel-example-text-editor-triggers',
+    shadow: true,
+    styleUrl: 'text-editor-custom-triggers.scss',
+})
+export class TextEditorCustomTriggersExample {
```
[the remark this answers is not in the document]
Comment (civing) [reply src/components/text-editor/examples/text-editor-custom-triggers.tsx:41]: @LucyChyzhova was adding a lot of good things but yeah, maybe we can simplify

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 10 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts:11]: I can change it in a follow up PR, there are some more things to do than simply changing the type to string

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 11 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: I'm not really sure we need to export it? See https://github.com/Lundalogik/lime-elements/pull/3268/commits/58d429f27b22620d17184f27041dc0c8f2bac6c6

Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: Ah, I should have looked more closely 👀 
I was referring to the create plugin function. 🙈

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 6 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: See https://github.com/Lundalogik/lime-elements/pull/3268/commits/0c355c81a6abaa86ba36262f648d16d93693539a 😅

File: src/components/text-editor/examples/text-editor-custom-triggers.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (LucyChyzhova) [reply src/components/text-editor/examples/text-editor-custom-triggers.tsx:41]: sure! will check what we can do

File: src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    transactions: Transaction[],
+    oldState: EditorState,
+): string => {
+    let textAdded = '';
+    let textRemoved = '';
+
+    transactions.forEach((transaction) => {
+        if (transaction.docChanged) {
+            transaction.steps.forEach((step) => {
+                if (
```
Comment (hannahu) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts:96]: Still some nested code, and also unnecessary inline code comments, but I'm ok merging it and get it in if it works as expected 😄

File: src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts (hunk 5 above)
Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts]: I will change the name of `CustomElement` to `CustomElementDefinition` and simply call the new on `CustomElement`

**Review by jgroth — APPROVED** [review]

File: src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+            stopTrigger();
+
+            return true;
+        }
+
+        if (isTrigger(event.key, triggerCharacters) && shouldTrigger(state)) {
+            activeTrigger = event.key;
+            triggerText = '';
+            triggerPosition = state.selection.$from.pos - triggerText.length;
+            sendTriggerEvent('triggerStart', view, activeTrigger, triggerText);
```
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts:158]: I came across this when working on https://github.com/Lundalogik/lime-elements/pull/3307. What's `triggerText` for, and is it supposed to ever be anything other than an empty string? If not, we can make the following change:

```diff
             triggerText = '';
-            triggerPosition = state.selection.$from.pos - triggerText.length;
+            triggerPosition = state.selection.$from.pos;
             sendTriggerEvent('triggerStart', view, activeTrigger, triggerText);
```

On the other hand, since it's a variable everywhere, it kinda looks like the intention is that it's not always going to remain an empty string? 🙂

Comment (civing) [reply src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts:158]: We used it on line 193 in this file. :) 

````ts
        if (updatedText !== triggerText) {
            triggerText = updatedText;
            sendTriggerEvent(
                'triggerChange',
                pluginView,
                activeTrigger,
                triggerText.slice(1),
            );
        }
````
